File: clojure/secret-handshake/src/secret_handshake.clj
(ns secret-handshake)

(defn too-big? [n]
  (>= n 32))

(defn commands [n]
  (if (too-big? n)
    (vector)
    (let [sorted? (>= n 16)]
      (loop [x (if sorted? (- n 16) n) events []]
        (cond 
          (zero? x) (if sorted? events (reverse events))
          :else
            (cond
              (>= x 8) (recur (- x 8) (conj events "jump"))
              (>= x 4) (recur (- x 4) (conj events "close your eyes"))
              (>= x 2) (recur (- x 2) (conj events "double blink"))
              (>= x 1) (recur (- x 1) (conj events "wink"))))))))

File: solutions/ruby/secret-handshake/2/secret_handshake.rb
class SecretHandshake
  ACTIONS = [
    'wink',            # 1 (00001)
    'double blink',    # 2 (00010)
    'close your eyes', # 4 (00100)
    'jump'             # 8 (01000)
  ]

  def initialize(mask)
    @mask = format('%05b', mask).to_i(2)
    @reverse = format('%05b', 16).to_i(2)
  end

  def commands
    enabled?(@reverse) ? actions.reverse : actions
  end

  private

  def actions
    @actions ||= ACTIONS.each_with_index.filter_map do |action, index|
      action if enabled?(1 << index)
    end
  end

  def enabled?(expr)
    (@mask & expr).nonzero?
  end
end
